[inductor] make mix order reduction work with dynamic shapes

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168209
* __->__ #168117


Internal models have dynamic shapes for rms/layer norm. Mix order reduction would be by-passed without this PR.

co-author this PR with Paul Zhang.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

Differential Revision: [D87378475](https://our.internmc.facebook.com/intern/diff/D87378475)

shunting314 committed:
[inductor] make mix order reduction work with dynamic shapes

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] make mix order reduction work with dynamic shapes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

shunting314 commented:
Need a bit more work to properly print dynamic shape expressions

PaulZhang12 commented:
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D87378475).

shunting314 committed:
Update on "[inductor] make mix order reduction work with dynamic shapes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [D87378475](https://our.internmc.facebook.com/intern/diff/D87378475)

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] make mix order reduction work with dynamic shapes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [D87378475](https://our.internmc.facebook.com/intern/diff/D87378475)

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] make mix order reduction work with dynamic shapes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [D87378475](https://our.internmc.facebook.com/intern/diff/D87378475)

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] make mix order reduction work with dynamic shapes"



Internal models have dynamic shapes for rms/layer norm. Mix order reduction would be by-passed without this PR.

co-author this PR with Paul Zhang.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [D87378475](https://our.internmc.facebook.com/intern/diff/D87378475)

[ghstack-poisoned]

PaulZhang12 commented:
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D87378475).

shunting314 committed:
Update on "[inductor] make mix order reduction work with dynamic shapes"



Internal models have dynamic shapes for rms/layer norm. Mix order reduction would be by-passed without this PR.

co-author this PR with Paul Zhang.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [D87378475](https://our.internmc.facebook.com/intern/diff/D87378475)

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] make mix order reduction work with dynamic shapes"



Internal models have dynamic shapes for rms/layer norm. Mix order reduction would be by-passed without this PR.

co-author this PR with Paul Zhang.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [D87378475](https://our.internmc.facebook.com/intern/diff/D87378475)

[ghstack-poisoned]

eellison commented on torch/_inductor/choices.py:
```diff
@@ -561,9 +561,11 @@ def can_fuse_horizontal(
-        if (
-            shared_data_score < config.score_fusion_memory_threshold
-        ) and not MixOrderReduction.can_fuse(node1, node2):
+        if MixOrderReduction.can_fuse(node1, node2):
```
for my knowledge - why do we not care about shared dat a?

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -253,19 +253,23 @@ def can_fuse(cls, node1: BaseSchedulerNode, node2: BaseSchedulerNode) -> bool:
+        # Call evaluate_expr rather than statically_known_geq since nrow can
+        # have dynamic shape in real models.
+        # Don't use hint directly since hint can be non-representative.
+        if not V.graph.sizevars.evaluate_expr(sympy.Ge(nrow * ncol, size_thres)):
```
similarly - wondering why not to guard?

shunting314 commented on torch/_inductor/codegen/simd.py:
```diff
@@ -1610,10 +1610,7 @@ def benchmark_codegened_module(
-            numel,
-            rnumel,
-        ):
+        if not V.graph.sizevars.evaluate_expr(sympy.Gt(numel, rnumel)):
```
I'm a bit conservative to adding guards. In real use cases we see, numel may have dynamic shape (e.g. jagged form of inputs), but rnumel is always static shape decided by model parameters
